srv/modules/runners/replace.py

```diff
@@ -0,0 +1,161 @@
+"""
+Runner to remove a single osd
+"""
+
+from __future__ import absolute_import
+from __future__ import print_function
+import time
+import logging
+import os
+# pylint: disable=import-error,3rd-party-module-not-gated,redefined-builtin
+import salt.client
+import salt.runner
+
+log = logging.getLogger(__name__)
+
+
+def help_():
+    """
+    Usage
+    """
+    usage = ('salt-run replace.osd id [id ...][force=True]:\n\n'
+             '    Removes an OSD from a minion\n'
+             '\n\n')
+    print(usage)
+    return ""
+
+
+def osd(*args, **kwargs):
+    """
+    Remove an OSD gracefully or forcefully on the minion
+
+    Note: If I were not in Salt, this whole routine would become a library.
+    This runner can be called by remove.osd and is only different by three
+    commands.  However, we have one runner call and multiple module calls below.
+    Trying to refactor these module calls into a __utils__ would require yet
+    another module.  Debugging all of this at runtime would be add to the fun,
+    so I didn't.
+    """
+    # Parameters for osd.remove module
+    supported = ['force', 'timeout', 'delay']
+    passed = ["{}={}".format(k, v) for k, v in kwargs.items() if k in supported]
+    log.debug("Converted kwargs: {}".format(passed))
+
+    # OSDs to remove
+    osds = list(str(arg) for arg in args)
+    if _checks_failed(osds, kwargs):
+        return ""
+
+    master_minion = _master_minion()
+
+    local = salt.client.LocalClient()
+    host_osds = local.cmd('I@roles:storage', 'osd.list', tgt_type='compound')
+
+    completed = osds
+    for osd_id in osds:
+        host = _find_host(osd_id, host_osds)
+        if host:
+            msg = _remove_minion(local, master_minion, osd_id, passed, host)
+            if msg:
+                print("{}\nFailed to remove osd {}".format(msg, osd_id))
+                completed.remove(osd_id)
+                continue
+
+            # Rename minion profile
+            minion_profile(host)
+
+    if 'called' in kwargs and kwargs['called']:
+        # Return for remove.osd
+        return {'master_minion': master_minion, 'osds': completed}
+    return ""
+
+
+def _checks_failed(osds, kwargs):
+    """
+    Check the safety, argument length.  Pause when multiple arguments are
+    passed to allow the admin to abort incorrect shell expansions
+    """
+    # Checks
+    if not __salt__['disengage.check']():
+        log.error('Safety engaged...run "salt-run disengage.safety"')
+        return True
+
+    if len(osds) < 1:
+        help_()
+        return True
+
+    if len(osds) > 1:
+        # Pause for a moment, let the admin see what they passed
+        print("Removing osds {} from minions\nPress Ctrl-C to abort".format(", ".join(osds)))
+        pause = 5
+        if 'pause' in kwargs and kwargs['pause']:
+            pause = kwargs['pause']
+        time.sleep(pause)
+
+    return False
+
+
+def _remove_minion(local, master_minion, osd_id, passed, host):
```

Comment:
the name of the function should be _remove_osd_from_minion or something in that direction.